src/Storages/Kafka/KafkaConsumer2.cpp

```diff
@@ -0,0 +1,375 @@
+#include <Storages/Kafka/KafkaConsumer2.h>
+
+#include <fmt/ranges.h>
+#include <cppkafka/exceptions.h>
+#include <cppkafka/topic_partition.h>
+#include <cppkafka/cppkafka.h>
+#include <cppkafka/topic_partition_list.h>
+#include <fmt/ostream.h>
+#include <boost/algorithm/string/join.hpp>
+
+#include <IO/ReadBufferFromMemory.h>
+#include <Storages/Kafka/StorageKafkaCommon.h>
+#include <Common/logger_useful.h>
+#include <Common/CurrentMetrics.h>
+#include <Common/ProfileEvents.h>
+
+#include <algorithm>
+#include <iterator>
+
+
+namespace CurrentMetrics
+{
+extern const Metric KafkaAssignedPartitions;
+extern const Metric KafkaConsumersWithAssignment;
+}
+
+namespace ProfileEvents
+{
+extern const Event KafkaRebalanceRevocations;
+extern const Event KafkaRebalanceAssignments;
+extern const Event KafkaRebalanceErrors;
+extern const Event KafkaMessagesPolled;
+extern const Event KafkaCommitFailures;
+extern const Event KafkaCommits;
+}
+
+namespace DB
+{
+
+using namespace std::chrono_literals;
+static constexpr auto EVENT_POLL_TIMEOUT = 50ms;
+static constexpr auto DRAIN_TIMEOUT_MS = 5000ms;
+
+
+bool KafkaConsumer2::TopicPartition::operator<(const TopicPartition & other) const
+{
+    return std::tie(topic, partition_id, offset) < std::tie(other.topic, other.partition_id, other.offset);
+}
+
+
+KafkaConsumer2::KafkaConsumer2(
+    ConsumerPtr consumer_,
+    LoggerPtr log_,
+    size_t max_batch_size,
+    size_t poll_timeout_,
+    const std::atomic<bool> & stopped_,
+    const Names & topics_)
+    : consumer(consumer_)
+    , log(log_)
+    , batch_size(max_batch_size)
+    , poll_timeout(poll_timeout_)
+    , stopped(stopped_)
+    , current(messages.begin())
+    , topics(topics_)
+{
+    // called (synchronously, during poll) when we enter the consumer group
+    consumer->set_assignment_callback(
```

Comment:
As I see we create num_consumers consumers in startup() synchronously, so I guess it is not good to do something heavy weight in constructor, but may be do this lazily? Plus at startup kafka might be unavailable, so we will not retry creation of consumers and can end up with zero created consumers

Reply:
Nice idea, changed the code.

Reply:
These functions only set the callbacks, so they cannot fail. I moved the `subscribe` call (which can actually fail) to `KafkaConsumer2::subscribeIfNotSubscribedYet`.